### ISSUE TITLE
OptionalUnit Rule on suspendCoroutine context

### ISSUE THREAD

**Issue by panoramix360** [detekt/detekt#4363]
## Expected Behavior
Don't throw an issue for OptionalUnit when used with suspendCoroutines that return nothing

In the following code, I can't remove Unit from the return because suspendCoroutine needs to known the function return type.
```
override suspend fun doAction(): Unit = suspendCoroutine { cont ->
        Integration.signOut() {
            it.fold(
                {
                    cont.resume(Unit)
                },
                { error ->
                    cont.resumeWithException(error as Exception)
                }
            )
        }
    }
```

## Observed Behavior
Detekt is complaining with a OptionalUnit rule.

## Steps to Reproduce
- Create a function that uses suspendCoroutine and returns nothing (Unit)

## Context
Detekt is not passing in this particular case in my project.

## Your Environment
* Version of detekt used: 1.18.1
* Version of Gradle used (if applicable): 7.0.3
* Operating System and version: Ubuntu 18.04

**Comment by severn-everett** [comment]
A workaround for this would be to specify the type in the generic bracket, e.g.
```
override suspend fun doAction() = suspendCoroutine<Unit> { cont ->
        Integration.signOut() {
            it.fold(
                {
                    cont.resume(Unit)
                },
                { error ->
                    cont.resumeWithException(error as Exception)
                }
            )
        }
    }
```

**Comment by panoramix360** [comment]
Hi @severn-everett, it worked like you suggested!

But I think that it's nice to address this on the other way around too.

Thank you for helping and opening a PR to adjust this.

**Comment by BraisGabin** [comment]
Thanks for the report @panoramix360. Closing because @severn-everett fixed it in #4371
